File: app/models/countdown_timer.rb
class CountdownTimer

  attr_accessor :delegate
  attr_reader :length, :remaining

  def self.sharedClient
    Dispatch.once { @instance ||= new }
    @instance
  end

  def initialize

    @remaining = 0
    @length = 0

  end

  def setTime(length)

    @length = length
    @remaining = length

  end

  def start

    unless @runTimer || @remaining.zero?
      @runTimer = NSTimer.scheduledTimerWithTimeInterval(1, target:self, selector:"tick", userInfo:nil, repeats:true)
    end

  end

  def stop

    @runTimer.invalidate if @runTimer
    @runTimer = nil

  end

  def reset

    stop
    @remaining = @length

    @delegate.countdownHasChanged

  end

  def active?

    !@runTimer.nil?

  end

  def tick

    @remaining -= 1
    @delegate.countdownHasChanged

    if @remaining == 0
      stop
      delegate.countdownHasFinished
    end

  end

  def addToRemaining(amount)

    @remaining += amount

    if @remaining < 0
      stop
      @remaining = 0
      @delegate.countdownHasChanged
      @delegate.countdownHasFinished
    else
      @delegate.countdownHasChanged
    end

  end

end

File: Countout/CountdownTimer.swift
import Foundation

class CountdownTimer: NSObject {
    
    static let sharedClient = CountdownTimer()
    
    var delegate: CountdownTimerDelegate?
    var length = 0 { didSet { remaining = length } }
    private(set) var remaining = 0
    var runTimer: NSTimer?
    
    func start() {
        if runTimer == nil && remaining != 0 {
            runTimer = NSTimer.scheduledTimerWithTimeInterval(1, target: self, selector: #selector(tick), userInfo: nil, repeats: true)
        }
    }
    
    func stop() {
        runTimer?.invalidate()
        runTimer = nil
    }
    
    func reset() {
        stop()
        remaining = length
        
        delegate?.countdownHasChanged()
    }
    
    func isActive() -> Bool {
        return runTimer != nil
    }
    
    func tick() {
        remaining -= 1
        delegate?.countdownHasChanged()
        
        if remaining == 0 {
            stop()
            delegate?.countdownHasFinished()
        }
    }
    
    func addToRemaining(amount: Int) {
        remaining += amount
        
        if remaining < 0 {
            stop()
            remaining = 0
            delegate?.countdownHasChanged()
            delegate?.countdownHasFinished()
        } else {
            delegate?.countdownHasChanged()
        }
    }
    
}

// MARK: CountdownTimerDelegate
protocol CountdownTimerDelegate {
    func countdownHasChanged()
    func countdownHasFinished()
}
